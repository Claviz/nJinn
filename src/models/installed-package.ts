export interface InstalledPackage {
    name: string;
    installed: string;
    latest: string;
}