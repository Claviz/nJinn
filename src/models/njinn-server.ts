export interface NjinnServer {
    close: () => Promise<void>;
}