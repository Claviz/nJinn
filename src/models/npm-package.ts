export interface NpmPackage {
    name: string;
    version: string;
}