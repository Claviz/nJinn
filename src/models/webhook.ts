export interface Webhook {
    url: string;
    headers: {
        [key: string]: string;
    }
}