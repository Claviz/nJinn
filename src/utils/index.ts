import execa from 'execa';

export async function getInstalledPackages(cwd: string): Promise<{ [key: string]: string }> {
    const npmListCommandResult = (await execa('npm', ['list', '--depth=0', '--json'], { cwd, })).stdout;
    if (npmListCommandResult) {
        const jsonResult = JSON.parse(npmListCommandResult);
        const npmPackageList = jsonResult.dependencies
        if (npmPackageList) {
            const packages: { [key: string]: string } = {};
            Object.keys(npmPackageList).forEach(function (key) {
                packages[key] = npmPackageList[key].version;
            });

            return packages;
        }
    }
    return {};
}