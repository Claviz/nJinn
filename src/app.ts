import Queue from 'bull';
import execa from 'execa';
import Fastify from 'fastify';
import fs from 'fs-extra';
import latestVersion from 'latest-version';
import path from 'path';
import { performance } from 'perf_hooks';
import rp from 'request-promise';

import { InstalledPackage } from './models/installed-package';
import { NjinnServer } from './models/njinn-server';
import { NpmPackage } from './models/npm-package';
import { ServerOptions } from './models/server-options';
import { Webhook } from './models/webhook';
import { getInstalledPackages } from './utils';

const REDIS_HOST = process.env.REDIS_HOST || 'localhost';

async function buildQueue(): Promise<Queue.Queue> {
    const queue = new Queue('nJinn', { redis: { host: REDIS_HOST } });
    queue.process(100, async function (job, done) {
        try {
            const { id, context } = job.data;
            const result = await require(path.resolve(`./generated/${id}`))(context);
            done(null, result);
        } catch (err) {
            done(err);
        }
    });

    queue.on('completed', async (job, result) => {
        await statusRequest(job.data.webhook, {
            ...job.data,
            status: 'success',
            result,
        });
    });

    queue.on('failed', async function (job, err) {
        await statusRequest(job.data.webhook, {
            ...job.data,
            status: 'fail',
        });
    })

    queue.on('active', async function (job, jobPromise) {
        await statusRequest(job.data.webhook, {
            ...job.data,
            status: 'loading',
        });
    })

    async function statusRequest(webhook: Webhook, data: any) {
        await rp({
            method: 'POST',
            uri: webhook.url,
            body: {
                id: data.id,
                result: data.result,
                status: data.status,
                timestamp: performance.now() + performance.timeOrigin,
            },
            rejectUnauthorized: false,
            json: true,
            headers: webhook.headers,
        })
    }

    await queue.isReady();

    return queue;
}

export async function startServer(options: ServerOptions): Promise<NjinnServer> {
    const fastify = Fastify();
    const queue = await buildQueue();

    fastify.post('/saveScript', async (request, reply) => {
        const { script, id } = request.body;
        if (!id) {
            throw new Error('id is required');
        }
        const fileName = `./generated/${id}`;
        await fs.outputFile(fileName, script);
        delete require.cache[path.resolve(fileName)];
        reply.send();
    });

    fastify.post('/queueJobs', (request, reply) => {
        const jobs = request.body as any[];
        if (jobs) {
            jobs.forEach(job => {
                queue.add(job, { removeOnComplete: true, removeOnFail: true, });
            });
        }
        reply.send();
    });

    fastify.post('/executeScript', async (request, reply) => {
        const { id, context } = request.body;
        const result = await require(path.resolve(`./generated/${id}`))(context);
        reply.send(result);
    });

    fastify.post('/installPackages', async (request, reply) => {
        await installPackages(request.body);
        reply.send();
    });

    async function installPackages(packages: NpmPackage[] = []) {
        await execa('npm', ['init', '-y'], { cwd: './generated' });
        const installedPackages = await getPackages(packages.map(x => x.package));
        for (let i = 0; i < packages.length; i++) {
            const packageName = packages[i].package;
            const version = packages[i].version || 'latest';
            const installedPackage = installedPackages.find(x => x.name === packageName);
            const packageWithVersion = `${packageName}@${version}`;
            const alreadyInstalled = installedPackage && (installedPackage.installed === version || (version === 'latest' && installedPackage.installed === installedPackage.latest));
            if (!alreadyInstalled) {
                await execa('npm', ['install', packageWithVersion], { cwd: './generated' });
            }
        }
    }

    async function getPackages(packageNames: string[] = []): Promise<InstalledPackage[]> {
        const packages = await getInstalledPackages('./generated');
        const installedPackages = [];
        for (let i = 0; i < packageNames.length; i++) {
            const packageName = packageNames[i];
            const latestPackageVersion = await latestVersion(packageName);
            if (packages[packageName]) {
                installedPackages.push({
                    name: packageName,
                    installed: packages[packageName],
                    latest: latestPackageVersion,
                });
            }
        }

        return installedPackages;
    }

    fastify.addHook('onClose', async (instance, done) => {
        await queue.close();
        done();
    });

    await fastify.listen(options.port, '0.0.0.0');
    console.log(`Server started at http://localhost:${options.port}`);

    return {
        close: fastify.close,
    };
}
