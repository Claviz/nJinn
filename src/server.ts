#!/usr/bin/env node
import commander from 'commander';

import { startServer } from './app';

(async () => {
    const program = new commander.Command();
    program
        .name('njinn')
        .requiredOption('-p, --port <port>', 'port');
    program.parse(process.argv);
    const { port } = program;
    await startServer({ port });
})();