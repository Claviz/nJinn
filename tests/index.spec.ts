import fs from 'fs-extra';
import { performance } from 'perf_hooks';
import rp from 'request-promise';

import { startServer } from '../src/app';
import { NpmPackage } from '../src/models/npm-package';
import { collectData } from './helpers';

import latestVersion from 'latest-version';
import { getInstalledPackages } from '../src/utils';
import { NjinnServer } from '../src/models/njinn-server';

jest.mock('perf_hooks');
let server: NjinnServer;

beforeEach(() => {
    jest.resetModules();
});

beforeAll(async () => {
    server = await startServer({ port: 3033 });
});

afterAll(async () => {
    await server.close();
});

beforeEach(async () => {
    await fs.emptyDir(`./generated`);
    let time = 0;
    (performance as any).timeOrigin = 0;
    performance.now = () => time++;
});

describe('/saveScipt', () => {
    it('saves script to the file system with correct file name', async () => {
        const script = `console.log('abc')`;
        await rp({
            method: 'POST',
            url: 'http://localhost:3033/saveScript',
            body: {
                id: 'abc.js',
                script,
            },
            json: true,
        });

        const fileContents = await fs.readFile('./generated/abc.js', 'utf8');
        expect(fileContents).toEqual(script);
    });

    it(`throws error if id is not specified`, async () => {
        try {
            await rp({
                method: 'POST',
                url: 'http://localhost:3033/saveScript',
                body: {},
                json: true,
            });
        } catch (e) {
            expect(e).toMatchSnapshot();
        }
    });
});

describe('/installPackages', () => {
    it('installs latest version of package if version is not specified', async () => {
        await rp({
            method: 'POST',
            url: 'http://localhost:3033/installPackages',
            body: [{
                name: 'is-number',
            }] as NpmPackage[],
            json: true,
        });
        const packages = await getInstalledPackages('./generated');
        const latestPackageVersion = await latestVersion('is-number');
        expect(packages['is-number']).toEqual(latestPackageVersion);
    });

    it('installs specific version of package', async () => {
        const version = '2.0.0';
        await rp({
            method: 'POST',
            url: 'http://localhost:3033/installPackages',
            body: [{
                name: 'is-number',
                version,
            }] as NpmPackage[],
            json: true,
        });
        const packages = await getInstalledPackages('./generated');
        expect(packages['is-number']).toEqual(version);
    });
});

describe('/queueJobs', () => {
    it('executes successfull script in the background and sends status', async () => {
        await rp({
            method: 'POST',
            url: 'http://localhost:3033/saveScript',
            body: {
                id: 'abc.js',
                script: `module.exports = () => { return 123; }`,
            },
            json: true,
        });
        const data = await collectData({
            port: 3034,
            length: 2,
            afterInit: async () => {
                await rp({
                    method: 'POST',
                    url: 'http://localhost:3033/queueJobs',
                    body: [{
                        id: 'abc.js',
                        webhook: { url: 'http://localhost:3034/webhook' }
                    }],
                    json: true,
                })
            }
        });
        expect(data).toMatchSnapshot();
    });

    it('executes failing script in the background and sends status', async () => {
        await rp({
            method: 'POST',
            url: 'http://localhost:3033/saveScript',
            body: {
                id: 'abc.js',
                script: `module.exports = () => { throw new Error(123); }`,
            },
            json: true,
        });
        const data = await collectData({
            port: 3034,
            length: 2,
            afterInit: async () => {
                await rp({
                    method: 'POST',
                    url: 'http://localhost:3033/queueJobs',
                    body: [{
                        id: 'abc.js',
                        webhook: { url: 'http://localhost:3034/webhook' }
                    }],
                    json: true,
                })
            }
        });
        expect(data).toMatchSnapshot();
    });
});

describe('/executeScript', () => {
    it('executes successfull script', async () => {
        await rp({
            method: 'POST',
            url: 'http://localhost:3033/saveScript',
            body: {
                id: 'abc.js',
                script: `module.exports = () => { return 123; }`,
            },
            json: true,
        });
        const result = await rp({
            method: 'POST',
            url: 'http://localhost:3033/executeScript',
            body: {
                id: 'abc.js',
            },
            json: true,
        });
        expect(result).toEqual(123);
    });

    it('executes failing script', async () => {
        await rp({
            method: 'POST',
            url: 'http://localhost:3033/saveScript',
            body: {
                id: 'abc.js',
                script: `module.exports = () => { throw new Error(123); }`,
            },
            json: true,
        });
        try {
            await rp({
                method: 'POST',
                url: 'http://localhost:3033/executeScript',
                body: {
                    id: 'abc.js',
                },
                json: true,
            });
        } catch (err) {
            // console.log(err);
            expect(err).toMatchSnapshot();
        }
    });
});