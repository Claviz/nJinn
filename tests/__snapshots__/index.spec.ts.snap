// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/executeScript executes failing script 1`] = `[StatusCodeError: 500 - {"statusCode":500,"error":"Internal Server Error","message":"123"}]`;

exports[`/queueJobs executes failing script in the background and sends status 1`] = `
Array [
  Object {
    "id": "abc.js",
    "status": "loading",
    "timestamp": 0,
  },
  Object {
    "id": "abc.js",
    "status": "fail",
    "timestamp": 1,
  },
]
`;

exports[`/queueJobs executes successfull script in the background and sends status 1`] = `
Array [
  Object {
    "id": "abc.js",
    "status": "loading",
    "timestamp": 0,
  },
  Object {
    "id": "abc.js",
    "result": 123,
    "status": "success",
    "timestamp": 1,
  },
]
`;

exports[`/saveScipt throws error if id is not specified 1`] = `[StatusCodeError: 500 - {"statusCode":500,"error":"Internal Server Error","message":"id is required"}]`;
